EDUH-2.1

DimaSzZ opened the pull request:
EDUH-2.1

MrazotaIzTirasa committed:
Поверхностная архитектура

MrazotaIzTirasa committed:
Eduh 2.1
Разработка:
Написал логику для репозиториев,написал сервисы, сделал автомаппер, сделал валидацию и экзепшены, создал IoC контейнер, написал интерфейс UoW

1nsanov commented on EduHub.sln:
```diff
@@ -14,6 +14,12 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{8DABF615
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Application", "Application", "{F2D0D35D-0072-46C5-B745-EC193360C9F4}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Infrastructure", "Infrastructure", "{B6FEDD40-9242-4A0A-97EF-16C433F2C264}"
```
Эта папка пуста в решении

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/AutoMapper/MappingProfile.cs:
```diff
@@ -0,0 +1,60 @@
+using EduHub.StudentService.Domain.Entities;
+using EduHub.StudentService.Domain.ValueObjects;
+
+namespace EduHub.StudentService.Application.Services.AutoMapper;
```
Папку достаточно назвать Mapping

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/AutoMapper/MappingProfile.cs:
```diff
@@ -0,0 +1,60 @@
+                ));
+        CreateMap<Educator, EducatorResponseDto>();
+        
+        CreateMap<CourseRequestDto, Course>()
```
Все мапинги в одном месте не есть хорошо, будет большей моделей и начнется каша, лучше всего под каждую сущность свой файл мапинг конфигурации

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Dto/Educator/EducatorRequestDto.cs:
```diff
@@ -0,0 +1,5 @@
+
+namespace EduHub.StudentService.Application.Services.Dto.Educator;
+
+public record EducatorRequestDto(string FirstName, string Surname, string Patronymic, Gender Gender, string Phone, int WorkExperience, DateOnly DateEmployment);
```
Если много аргументов, лучше всего писать каждый с новый строки (с столбик) для лучшей читаемости. По правилу примерно в одну строку можно уместить 4-5 аргументов, если больше, то переносим на новую строку

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Dto/Educator/EducatorResponseDto.cs:
```diff
@@ -0,0 +1,6 @@
+
+namespace EduHub.StudentService.Application.Services.Dto.Educator;
+
+public record EducatorResponseDto(Guid Id,FullName FullName, Gender Gender, Phone Phone, int WorkExperience, DateOnly DateEmployment);
```
Между запятой и FullName нет пробела, модели (VO) для DTO не используем. В таких случаях пишем отдельными аргументами (FirstName, Surname и т.д.), или можешь написать отдельные DTO для FullName и Адреса, для телефона и почты нет необходимости DTO писать, достаточно string передавать. Исправить во всез DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Realization/EntityConflictException.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using EduHub.StudentService.Application.Services.Exceptions.Base;
+
+namespace EduHub.StudentService.Application.Services.Exceptions.Realization;
```
Итоговые классы исключений можно не выносить в отдельную папку

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/ExceptionsInfo.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace EduHub.StudentService.Application.Services.Exceptions;
+
+public static class ExceptionsInfo
```
Класс лучше назвать иначе, например ErrorMessages. И вынести в отдельную папку, например для таких целей часто используют папку с наименованием 'Primitives'

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/ExceptionsInfo.cs:
```diff
@@ -0,0 +1,7 @@
+
+public static class ExceptionsInfo
+{
+    public static string NotFound = "Resource cannot be found";
```
Rider подсказывает, что можно сделать readonly, а еще лучше - const

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/ExceptionsInfo.cs:
```diff
@@ -0,0 +1,7 @@
+public static class ExceptionsInfo
+{
+    public static string NotFound = "Resource cannot be found";
+    public static string Conflict = "Resource is Conflicted";
```
Сообщения об ошибка несут недостаточно информации. Какой ресурс не был найден, по какому значению и т.д. Должно быть что-то по типу 'Student with id '123' not found.' Подумай, как лучше реализовать, чтобы для каждой модели не писать отдельный ошибки на NotFound и Conflict. Можешь попробовать через string.Format

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/IUnitOfWork.cs:
```diff
@@ -0,0 +1,11 @@
+    ICourseRepository CoursesRepository { get; }
+    IEducatorRepository EducatorRepository { get; }
+    IEnrollmentRepository EnrollmentRepository { get; }
+    IStudentRepository StudentRepository { get; }
```
А что, если на понадобится применить интерфейс IUnitOfWork для другого дб контекста? Эти репозитории тогда уже не будут актуальны. Надо продумать данный момент, например есть самый базовый интерфейс UoW с описанием методов. Далее от него можно создать уже более специфичный интерфейс UoW конректно для данного дб контекста (сервиса условно). И в нем уже можно описать свойства для репозиториев

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/IUnitOfWork.cs:
```diff
@@ -0,0 +1,11 @@
+    IEnrollmentRepository EnrollmentRepository { get; }
+    IStudentRepository StudentRepository { get; }
+    
+    Task<int> SaveChanges();
```
Все асинхронные методы принято называть с суфиксом Async. Например SaveChangesAsync и т.д. Исправить везде.

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IBaseRepository.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace EduHub.StudentService.Application.Services.Interfaces;
+
+public interface IBaseRepository<T> where T : BaseEntity
```
1. Для репозиториев лучше создать отдельную папку в папке Interfaces. И в ней еще создать папку Core, и в неё же пихать базовую репу. А остальные репы в папку Interfaces/Repositories
2. Дженерик тип надо называть более осмыслено, а не просто - T

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IBaseRepository.cs:
```diff
@@ -0,0 +1,12 @@
+    Task<T> Add(T entity, CancellationToken cancellationToken);
+    Task<T> Update(T entity, CancellationToken cancellationToken);
+    
+    Task<List<T>> GetAllEntities(CancellationToken cancellationToken);
```
Достаточно просто GetAll

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IBaseRepository.cs:
```diff
@@ -0,0 +1,12 @@
+    Task<T> Update(T entity, CancellationToken cancellationToken);
+    
+    Task<List<T>> GetAllEntities(CancellationToken cancellationToken);
+    Task<T> Delete(Guid id,CancellationToken cancellationToken);
```
1. Уверен, что метод удаления должен принимать только Id?
2. Я не вижу метода для получения сущности по Id. И тут не хватает пробела между , и CancellationToken.

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/ICourseRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface ICourseRepository: IBaseRepository<Course>
+{
+    Task<Course> GetCourseById(Guid id, CancellationToken cancellationToken);
```
Зачем, если можно функционал получения сущности по Id поместить в базовую репу. Ведь у каждой сущности будет одна и та же логика

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IEducatorRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface IEducatorRepository : IBaseRepository<Educator>
+{
+    Task<Educator> GeEducatorById(Guid id,CancellationToken cancellationToken);
```
Тоже самое, что и в ICourseRepository

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IEnrollmentRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface IEnrollmentRepository : IBaseRepository<Enrollment>
+{
+    Task<List<Enrollment>> GetStudentEnrollments(Student student,CancellationToken cancellationToken);
```
Лучше использовать интерфейс IList. Из названия метода не до конца понятно, что он делает и как. лучше будет GetListByStudentIdAsync например. И зачем ты передаешь целую модель студента, если тебе нужен только его Id ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IStudentRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface IStudentRepository : IBaseRepository<Student>
+{
+    Task<Student> GetStudentById(Guid id, CancellationToken cancellationToken);
```
Тоже самое, что и в ICourseRepository

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Ioc/DependencyInjection.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace EduHub.StudentService.Application.Services.Ioc;
+
+public static class DependencyInjection
```
А зачем тебе сейчас DI ? + Название папки мне не нравится, IoC это немного не то

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace EduHub.StudentService.Application.Services.Services;
+
+public class CourseService
```
Интерфейсы сервисов потерял по дороге, относится ко всем

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    
+    public CourseService(IMapper mapper, IUnitOfWork unitOfWork)
+    {
+        _mapper = mapper;
```
А если вдруг с DI приключилась шляпа и mapper = null, и в методе будет обращение _mapper, который null, выбросит ошибку NullReference. Лучше такое проверять Guard'ом на null. Например _mapper = Guard.Against.Null(mapper). Исправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        _unitOfWork = unitOfWork;
+    }
+    
+    public async Task AddNewCourse(CourseRequestDto courseDto, CancellationToken cancellationToken)
```
А так понятно, что добавляем новый объект, и понятно что курс, исходя и местоположения в CourseService. Достаточно назвать AddAsync или CreateAsync

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    
+    public async Task AddNewCourse(CourseRequestDto courseDto, CancellationToken cancellationToken)
+    {
+        await new CourseRequestDtoValidator().ValidateAndThrowAsync(courseDto, cancellationToken);
```
Если courseDto будет null? Такое надо проверять гуардом до бизнес валидации. Исправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        
+        var course = _mapper.Map<CourseRequestDto, Course>(courseDto);
+        var existingCourse = await _unitOfWork.CoursesRepository.Add(course, cancellationToken);
+        if (existingCourse != null)
```
Обговаривали в телеге. Исправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        if (existingCourse != null)
+            throw new EntityConflictException<Course>();
+        
+        await _unitOfWork.SaveChanges();
```
Метод по добавлению должен вовзращать Дто модель, либо Id модели. Тут сам думай, как лучше

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        if (existingCourse == null)
+            throw new EntityNotFoundException<Course>();
+        
+        await _unitOfWork.SaveChanges();
```
Метод по добавлению должен вовзращать Дто модель, либо Id модели. Тут сам думай, как лучше

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        await _unitOfWork.SaveChanges();
+    }
+    
+    public async Task UpdateCourse(CourseRequestDto courseDto, CancellationToken cancellationToken)
```
Можно просто UpdateAsync

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        await new CourseRequestDtoValidator().ValidateAndThrowAsync(courseDto, cancellationToken);
+        
+        var course = _mapper.Map<CourseRequestDto, Course>(courseDto);
+        var existingCourse = await _unitOfWork.CoursesRepository.Update(course, cancellationToken);
```
1. Как ты собрался делать Update, когда в модели Дто не передается Id?)
2. Обновляем сущности через метод Update (Который писал в моделях)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    
+    public async Task DeleteCourse(Guid id, CancellationToken cancellationToken)
+    {
+        await _unitOfWork.CoursesRepository.Delete(id, cancellationToken);
```
А если по такому Id нет записи в БД? Возможно, надо сначала получить объект, потом проверить его на null, выбросить соотв. исключение по необходимости и только потом удалить объект.

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    public async Task<List<string>> GetListCourses(CancellationToken cancellationToken)
+    {
+        var courses = await _unitOfWork.CoursesRepository.GetAllEntities(cancellationToken);
+        if (courses == null)
```
Уверен, что метод может тебе вернуть null?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        if (courses == null)
+            throw new EntityNotFoundException<Course>();
+        
+        return _mapper.Map<List<CourseResponseDto>>(courses).Select(course => course.Name).ToList();
```
нононо, надо возвращать список DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        return _mapper.Map<List<CourseResponseDto>>(courses).Select(course => course.Name).ToList();
+    }
+    
+    public async Task<CourseResponseDto> GetInfoCourse(Guid id, CancellationToken cancellationToken)
```
Достаточно просто GetByIdAsync

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    
+    public async Task<CourseResponseDto> GetInfoCourse(Guid id, CancellationToken cancellationToken)
+    {
+        var course = await _unitOfWork.CoursesRepository.GetCourseById(id, cancellationToken);
```
Если id будет дэфолтным значением, зачем вызывать метод репы и обращаться к Бд? Лучше сначала проверить через Guard Id

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+    {
+        var course = await _unitOfWork.CoursesRepository.GetCourseById(id, cancellationToken);
+        if (course == null)
+            throw new EntityNotFoundException<Course>();
```
Инлайн стараемся не использовать, лучше через фигурные скобки для читаемости

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,70 @@
+        await _unitOfWork.SaveChanges();
+    }
+    
+    public async Task<List<string>> GetListCourses(CancellationToken cancellationToken)
```
Вместо List лучше будет IList

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Realization/EntityNotFoundException.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace EduHub.StudentService.Application.Services.Exceptions.Realization;
+
+public class EntityNotFoundException<T> : BaseNotFoundException
```
Дженерик тип надо более понятно называть, а не просто - T. И хорошо ли, что в дженерик тип можно подставить любой тип?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/CourseRequestDtoValidator.cs:
```diff
@@ -0,0 +1,13 @@
+    public CourseRequestDtoValidator()
+    {
+        RuleFor(dto => dto.Name)
+            .MaximumLength(50);
```
А если строка будет пустая?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/CourseRequestDtoValidator.cs:
```diff
@@ -0,0 +1,13 @@
+
+public class CourseRequestDtoValidator : AbstractValidator<CourseRequestDto>
+{
+    public CourseRequestDtoValidator()
```
Нет валидации на дэфолтное значение EducatorId

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/EducatorRequestDtoValidator.cs:
```diff
@@ -0,0 +1,35 @@
+    public EducatorRequestDtoValidator()
+    {
+        RuleFor(dto => dto.FirstName)
+            .NotEmpty().WithMessage("FirstName is required")
```
Сообщения ошибок не хардкодим, выносим куда-нибудь

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/EducatorRequestDtoValidator.cs:
```diff
@@ -0,0 +1,35 @@
+            .NotEmpty().WithMessage("FirstName is required")
+            .MinimumLength(2).WithMessage("FirstName must be at least 2 characters long")
+            .MaximumLength(60).WithMessage("FirstName cannot exceed 60 characters")
+            .Matches("^[A-Za-z]+$").WithMessage("FirstName must contain only letters");
```
Паттерны регулярок выносим в отельное место. А еще твоя регулярка залупа, не пропускает кириллицу и т.д. Подумай еще раз над ней

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/EducatorRequestDtoValidator.cs:
```diff
@@ -0,0 +1,35 @@
+        
+        RuleFor(dto => dto.Phone)
+            .NotEmpty().WithMessage("Phone number is required")
+            .Matches(@"^\+373(?:6|7)[0-9]{7}$").WithMessage("Invalid phone number format for PMR");
```
Проверь еще раз формат ПМР номеров, регулярка не дописана.

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/EducatorRequestDtoValidator.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class EducatorRequestDtoValidator : AbstractValidator<EducatorRequestDto>
+{
+    public EducatorRequestDtoValidator()
```
DateEmployment, Gender проверять не надо?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/StudentRequestDtoValidator.cs:
```diff
@@ -0,0 +1,56 @@
+        RuleFor(dto => dto.Email)
+            .NotEmpty().WithMessage("Email is required")
+            .MaximumLength(255).WithMessage("Email length cannot exceed 255 characters")
+            .Must(email => email.Contains("@")).WithMessage("Invalid email format");
```
Хреновая проверка почты, нужна регулярка

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/StudentRequestDtoValidator.cs:
```diff
@@ -0,0 +1,56 @@
+        
+        RuleFor(dto => dto.Avatar)
+            .NotEmpty().WithMessage("Avatar URL is required")
+            .Matches(@"\.(png|jpg)$").WithMessage("Avatar URL must be a .png or .jpg image");
```
Не уверен в правильности регулярки на url аватара

1nsanov commented:
Все комментарии в CourseService относятся ко ВСЕМ остальным сервисам

MrazotaIzTirasa committed:
Eduh 2.1
Разработка и правка
Переделал валидаторы, исправил сервисы, дописал UoW, исправил Dto, в общем все, что было в гит хабе

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Dto/Course/CourseRequestDto.cs:
```diff
@@ -0,0 +1,3 @@
+﻿namespace EduHub.StudentService.Application.Services.Dto.Course;
+
+public record CourseRequestDto(string Name, string Description, Guid EducatorId);
```
Раз уже пишешь полные имена для дто, понятней будет CourseRequestCreateDto. А лучше уже CourseCreateDto, CourseUpdateDto, CourseDto. И так со всеми DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Dto/Educator/EducatorResponseDto.cs:
```diff
@@ -0,0 +1,14 @@
+
+public record EducatorResponseDto(
+    Guid Id,
+    string FirsName,
```
Опечатка в наименовании аргумента

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Dto/Student/StudentResponseDto.cs:
```diff
@@ -0,0 +1,19 @@
+    string Patronymic,
+    Gender Gender,
+    DateOnly DateBirth,
+    Email Email,
```
VO убрать из DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Primitives/ErrorMessages.cs:
```diff
@@ -0,0 +1,16 @@
+
+public static class ErrorMessages
+{
+    public const string NotFound = "{0}  not found.";
```
Есть лишний пробел. В ошибку можно еще добавить по какому полю и какому значение приозошла ошибка

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Primitives/ErrorMessages.cs:
```diff
@@ -0,0 +1,16 @@
+public static class ErrorMessages
+{
+    public const string NotFound = "{0}  not found.";
+    public const string Conflict = "{0} with id '{1}' is conflicted.";
```
Возомжно эта ошибка пригодится не только для Id

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Regular/RegexPatterns.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace EduHub.StudentService.Application.Services.Exceptions.Regular;
```
Папку Primitives переместить в корень проекта и в неё пихнуть ErrorMessages и RegexPatterns

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Regular/RegexPatterns.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public const string NamePattern = @"^[\p{L}]+$";
+    public const string PmrPhonePattern = @"^\+373(?:2\d{7}|[67]\d{7})$";
+    public const string EmailPattern = @"^[^@\s]+@[^@\s]+\.[^@\s]+$";
```
Проверял регулярки?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Regular/RegexPatterns.cs:
```diff
@@ -0,0 +1,9 @@
+    public const string NamePattern = @"^[\p{L}]+$";
+    public const string PmrPhonePattern = @"^\+373(?:2\d{7}|[67]\d{7})$";
+    public const string EmailPattern = @"^[^@\s]+@[^@\s]+\.[^@\s]+$";
+    public const string FilePngJpgPattern = @"\.(png|jpg";
```
Не вижу проверки на http(s)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/Base/BaseConflictException.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace EduHub.StudentService.Application.Services.Exceptions.Base;
```
В проекте Domain базовые классы были в папке под названием Core, а тут Base. Определись с одним из двух названий и используй везде одно

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Repositories/ICourseRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace EduHub.StudentService.Application.Services.Interfaces.Repositories;
+
+public interface ICourseRepository: IBaseRepository<Course>
```
Нет пробела между ICourseRepository и двоеточием

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Repositories/ICourseRepository.cs:
```diff
@@ -0,0 +1,8 @@
+namespace EduHub.StudentService.Application.Services.Interfaces.Repositories;
+
+public interface ICourseRepository: IBaseRepository<Course>
+{
```
Rider подсказывает, что можно удалить фигурные скобки. В других репах тоже есть

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/Core/IUnitOfWork.cs:
```diff
@@ -0,0 +1,6 @@
+
+public interface IUnitOfWork : IDisposable
+{
+    Task<int> SaveChangesAsync();
```
Не вижу токена отмены

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/IPgUnitOfWork.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace EduHub.StudentService.Application.Services.Interfaces.UoW;
+
+public interface IPgUnitOfWork : IUnitOfWork
```
Название залупа, лучше уже назвать IStudentUnitOfWork

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IEducatorService.cs:
```diff
@@ -0,0 +1,16 @@
+
+public interface IEducatorService
+{
+    public Task<EducatorResponseDto> AddNewAsync(EducatorRequestDto educatorDto, CancellationToken cancellationToken);
```
Можно просто 'AddAsync'

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/ICourseService.cs:
```diff
@@ -0,0 +1,13 @@
+    public Task DeleteAsync(Guid id, CancellationToken cancellationToken);
+    public Task<IList<CourseResponseDto>> GetListAsync(CancellationToken cancellationToken);
+    
+    public Task<CourseResponseDto> GetBuIdAsync(Guid id, CancellationToken cancellationToken);
```
Ошибка в названии метода

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IEducatorService.cs:
```diff
@@ -0,0 +1,16 @@
+    
+    public Task<List<(string, string, string)>> GetListAsync(CancellationToken cancellationToken);
+    
+    public Task<EducatorResponseDto> GetInfoAsync(Guid id, CancellationToken cancellationToken);
```
Называй уже везде метод для получение объекта по Id в сервисах GetByIdAsync

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/IEnrollmentService.cs:
```diff
@@ -0,0 +1,11 @@
+
+public interface IEnrollmentService
+{
+    public Task<EnrollmentResponseDto> AddStudentToCourseAsync(EnrollmentRequestDto enrollmentRequestDto, CancellationToken cancellationToken);
```
Достаточно про AddAsync

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,83 @@
+        return _mapper.Map<List<EnrollmentResponseDto>>(enrollments);
+    }
+    
+    public async Task<List<Guid>> GetListStudentsAsync(CancellationToken cancellationToken)
```
Возвращай DTO, а не список Guid

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,106 @@
+        await _unitOfWork.SaveChangesAsync();
+    }
+    
+    public async Task<List<(string, string, string)>> GetListAsync(CancellationToken cancellationToken)
```
Че за хуйня с кортежами? Вовзращай нормальные человеческие DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/StudentService.cs:
```diff
@@ -0,0 +1,105 @@
+        await _unitOfWork.SaveChangesAsync();
+    }
+    
+    public async Task<List<(string, string, string)>> GetListAsync(CancellationToken cancellationToken)
```
Че за хуйня с кортежами? Вовзращай нормальные человеческие DTO

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/Educator/EducatorMappingProfile.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace EduHub.StudentService.Application.Services.Mapping.Educator
+{
+    public class EducatorMappingProfile : Profile
```
Зачем папки для каждого класса ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/Course/CourseMappingProfile.cs:
```diff
@@ -0,0 +1,24 @@
+                        dto.EducatorId
+                    ));
+            
+            CreateMap<CourseRequestUpdateDto, Domain.Entities.Course>().ReverseMap();
```
Не используй ReverseMap

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/Student/StudentMappingProfile.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        public StudentMappingProfile()
+        {
+            CreateMap<Domain.Entities.Student, StudentResponseDto>().ReverseMap();
```
Неполная кфг мапера, он не настолько умный, чтобы мапить VO в отдельные поля. И это не единственное такое место

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/Student/StudentMappingProfile.cs:
```diff
@@ -0,0 +1,34 @@
+                .ForPath(dest => dest.FullName.Patronymic, opt => opt.MapFrom(src => src.Patronymic))
+                .ForPath(dest => dest.Address.City, opt => opt.MapFrom(src => src.City))
+                .ForPath(dest => dest.Address.Street, opt => opt.MapFrom(src => src.Street))
+                .ForPath(dest => dest.Address.NumberHouse, opt => opt.MapFrom(src => src.NumberHouse));
```
Зачем ForPath'ы ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+        }
+        
+        dbCourse.Update(course.Name,course.Description,course.EducatorId);
+        await _unitOfWork.CoursesRepository.UpdateAsync(dbCourse, cancellationToken);
```
Можно не вызывать Update, объект dbCourse уже и так отслеживается дб контекстом (так как трекинг по умолчанию включен). Исправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+        
+        await new CourseRequestUpdateDtoValidator().ValidateAndThrowAsync(courseDto, cancellationToken);
+        
+        var course = _mapper.Map<CourseRequestUpdateDto, Course>(courseDto);
```
Зачем тебе здесь мапинг, если снизу все равно отдельно каждое поле передаешь?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+        
+        var course = _mapper.Map<CourseRequestUpdateDto, Course>(courseDto);
+        
+        var dbCourse = await _unitOfWork.CoursesRepository.GetByIdAsync(courseDto.Id,cancellationToken);
```
Можно просто course

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+            throw new EntityNotFoundException<Course>();
+        }
+        
+        dbCourse.Update(course.Name,course.Description,course.EducatorId);
```
Нет пробелов между аргументами

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+        var dbCourse = await _unitOfWork.CoursesRepository.GetByIdAsync(id, cancellationToken);
+        if (dbCourse == null)
+        {
+            throw new EntityNotFoundException<Course>();
```
Блок с получением и выбросом ошибки уже повторяется дважды, можно вынести в приватный метод

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+        var courses = await _unitOfWork.CoursesRepository.GetAllAsync(cancellationToken);
+        if (courses == null)
+        {
+            throw new EntityNotFoundException<Course>();   
```
Еще раз, зачем ты проверяешь здесь на null и кидаешь ошибку?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,106 @@
+        var educator = _mapper.Map<EducatorRequestDto, Educator>(educatorDto);
+        await _unitOfWork.EducatorRepository.AddAsync(educator, cancellationToken);
+       
+        
```
Лишняя пустая строка

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,106 @@
+        }
+        
+        dbEducator.Update(
+            new FullName(educator.FullName.FirstName,educator.FullName.Surname,educator.FullName.Patronymic),
```
Опять нет пробелов

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,106 @@
+            new Phone(educator.Phone.Value),
+            educator.YearsExperience,
+            educator.DateEmployment
+            );
```
Поправь форматирование ');'

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,106 @@
+        
+        await new EducatorRequestUpdateDtoValidator().ValidateAndThrowAsync(educatorDto, cancellationToken);
+        
+        var educator = _mapper.Map<EducatorRequestUpdateDto, Educator>(educatorDto);
```
Не вижу смысла в маппинге

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,97 @@
+    
+    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken)
+    {
+        Guard.Against.Null(id);
```
Зачем проверять на Null, если Guid не может быть null? )0) Поправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/ICourseService.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace EduHub.StudentService.Application.Services.Interfaces;
+
+public interface ICourseService
```
Вынести интерфейсы сервисов в отдельную папку

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,83 @@
+    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken)
+    {
+        Guard.Against.Null(id);
+        var dbEnrollment = await _unitOfWork.EnrollmentRepository.GetByIdAsync(id,cancellationToken);
```
Можно и без db в названии переменной

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/StudentService.cs:
```diff
@@ -0,0 +1,105 @@
+        
+        await new StudentRequestUpdateDtoValidator().ValidateAndThrowAsync(studentDto, cancellationToken);
+        
+        var student = _mapper.Map<StudentRequestUpdateDto, Student>(studentDto);
```
Мапинг лишний

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/StudentService.cs:
```diff
@@ -0,0 +1,105 @@
+            new Email(studentDto.Email),
+            new Phone(studentDto.Phone),
+            new Address(studentDto.City,studentDto.Street,studentDto.NumberHouse)
+            );
```
Поправить форматирование

MrazotaIzTirasa committed:
EDUH 2.1.
Разработка. Документирование проекта
Исправил все примечаня

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Exceptions/EntityNotFoundException.cs:
```diff
@@ -0,0 +1,21 @@
+    /// </summary>
+    /// <param name="field">не найденное название поля</param>
+    /// <param name="value">не найденный объект объект</param>
+    public EntityNotFoundException(string field, object value) :  base(string.Format(ErrorMessages.NotFound, typeof(TEntity).Name, field, value))
```
field как по мне не совсем подходит, обычно такого рода поля называем paramName или parameterName (так называны аргументы в Guard методах). Можешь сделать глобальную замену через Ctrl + Shift + H

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Repositories/Core/IBaseRepository.cs:
```diff
@@ -0,0 +1,49 @@
+    /// </summary>
+    /// <param name="cancellationToken">токен отмены</param>
+    /// <returns>Возвращение всех объектов бд</returns>
+    Task<List<TEntity>> GetAllAsync(CancellationToken cancellationToken);
```
Тут тоже давай IList возвращать. В IEnrollmentRepository кстати IList возвращается

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/Core/IUnitOfWork.cs:
```diff
@@ -0,0 +1,24 @@
+namespace EduHub.StudentService.Application.Services.Interfaces.UoW.Core;
+
+/// <summary>
+/// Базовый репозиторий, являющийся родителем для всех реализаций UoW
```
Поправь комент, это не базовая репа)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/Core/IUnitOfWork.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Сохраняет изменение в бд
+    /// </summary>
+    /// <param name="cancellationToken">тоен отмены</param>
```
тоен )

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/IStudentUnitOfWork.cs:
```diff
@@ -0,0 +1,16 @@
+namespace EduHub.StudentService.Application.Services.Interfaces.UoW;
+
+/// <summary>
+/// Реализация UoW, которую я везде использую
```
Интересный коммент, но ладно

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/IStudentUnitOfWork.cs:
```diff
@@ -0,0 +1,16 @@
+    IEnrollmentRepository EnrollmentRepository { get; }
+    IStudentRepository StudentRepository { get; }
+    
+}
```
Нажмешь тут ctrl alt enter )0)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/CourseMappingProfile.cs:
```diff
@@ -0,0 +1,39 @@
+        /// </summary>
+        public CourseMappingProfile()
+        {
+            CreateMap<CourseCreateDto, Domain.Entities.Course>()
```
Domain.Entities. можно убрать, поправивь во всех мапингах

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/CourseMappingProfile.cs:
```diff
@@ -0,0 +1,39 @@
+                    ));
+            
+            CreateMap<CourseUpdateDto, Domain.Entities.Course>()
+                .ForMember(dest => dest.Name, opt => opt.MapFrom(src => src.Name))
```
Если название поля в двух моделях одинаковое, ForMember можно не писать. Исправить во всех маперах

1nsanov replied:
Почему из dto в domain не мапишь, как в случае с CourseCreateDto -> Course? При таком мапинге пропускается системная валидация в контсрукторе, написанная тобой. Исправить во всех таких же случаях

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/EducatorMappingProfile.cs:
```diff
@@ -0,0 +1,47 @@
+                .ForMember(dest => dest.Surname, opt => opt.MapFrom(src => src.FullName.Surname))
+                .ForMember(dest => dest.Patronymic, opt => opt.MapFrom(src => src.FullName.Patronymic))
+                .ForMember(dest => dest.Phone, opt => opt.MapFrom(src => src.Phone.Value))
+                .ForMember(dest => dest.WorkExperience, opt => opt.MapFrom(src => src.YearsExperience))
```
А не проще назвать в Domain и dto модели поле одинаково и не писать ForMember?)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Primitives/RegexPatterns.cs:
```diff
@@ -0,0 +1,9 @@
+
+public static class RegexPatterns
+{
+    public const string NamePattern = @"^[\p{L}\s'-]+$";
```
Не хватает комментариев. Не совсем понятно, что за Name, вдруг это наименование курса и т.д.

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/ValidatorExtension.cs:
```diff
@@ -0,0 +1,157 @@
+    public static IRuleBuilderOptions<TEntity, string> PersonName<TEntity>(this IRuleBuilder<TEntity, string> ruleBuilder)
+    {
+        return ruleBuilder
+            .NotEmpty().WithMessage(string.Format(ErrorMessages.Required, nameof(TEntity)))
```
Исходя из nameof(TEntity), я скажу что сообщение ошибки будет такого рода : "CourseUpdateDto is required". Уверен в корректности сообщения ошибок?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/ValidatorExtension.cs:
```diff
@@ -0,0 +1,157 @@
+    /// <param name="ruleBuilder"> свойство, хранящее объект и тип данных для проверки</param>
+    /// <typeparam name="TEntity"> объект, который будет валидироваться</typeparam>
+    /// <returns>Возвращает отвалидированный объект</returns>
+    public static IRuleBuilderOptions<TEntity, string> PersonName<TEntity>(this IRuleBuilder<TEntity, string> ruleBuilder)
```
Название дженерика достаточно T, тут по факту может быть почти все что угодно

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/ValidatorExtension.cs:
```diff
@@ -0,0 +1,157 @@
+    }
+    
+    /// <summary>
+    /// Проверка на валидность названия
```
Проверка названия на что? Тоже относится и к названию метода

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/ValidatorExtension.cs:
```diff
@@ -0,0 +1,157 @@
+    {
+        return ruleBuilder
+            .NotEmpty().WithMessage(string.Format(ErrorMessages.Required, nameof(TEntity)))
+            .Must(date => date <= DateOnly.FromDateTime(DateTime.Today)).WithMessage(ErrorMessages.BadDate);
```
Есть для меньше/больше спец. методы у валидатора

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/ValidatorExtension.cs:
```diff
@@ -0,0 +1,157 @@
+    public static IRuleBuilderOptions<TEntity, string> Name<TEntity>(this IRuleBuilder<TEntity, string> ruleBuilder)
+    {
+        return ruleBuilder
+            .Matches(RegexPatterns.NamePattern).WithMessage(string.Format(ErrorMessages.BadName, nameof(TEntity)))
```
Проверка по регулярки должна быть в самом конце, она же самая "тяжелая" и еще много причин, почему она не должны быть первой

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Validations/Educator/EducatorCreateDtoValidator.cs:
```diff
@@ -0,0 +1,34 @@
+namespace EduHub.StudentService.Application.Services.Validations.Educator;
+
+/// <summary>
+/// Вадитаор на EducatorCreateDto
```
Вадитаор )0)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Services/ServiceBase.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <summary>
+    /// Базовый класс сервисов, созданный для удобного пользования
+    /// </summary>
+    public abstract class ServiceBase
```
Этот класс не должен находится в папке с интерфейсами

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Services/ServiceBase.cs:
```diff
@@ -0,0 +1,41 @@
+        }
+        
+        /// <summary>
+        /// Универсальнй метод для получения id, в зависимости от типа репозитория
```
*Для получения Domain сущности

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/UoW/Core/IUnitOfWork.cs:
```diff
@@ -0,0 +1,24 @@
+    /// </summary>
+    /// <typeparam name="TRepository">отвечает за то, какой репозиторий вернуть</typeparam>
+    /// <returns>возвращает определенный репозиторий</returns>
+    IBaseRepository<TRepository> GetRepository<TRepository>() where  TRepository : BaseEntity;
```
TRepository и она ограничена BaseEntity? Что-то тут явно не так

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Services/ServiceBase.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <exception cref="EntityNotFoundException{TEntity}"></exception>
+        protected async Task<TEntity> GetEntityByIdAsync<TEntity>(Guid id, string field, CancellationToken cancellationToken) where TEntity : BaseEntity
+        {
+            var repository = _unitOfWork.GetRepository<TEntity>();
```
Так так так, а ты уверен, что указав Domain класс ты получишь нужную репу? Может тебе надо здесь указывать сам интерфейс репы?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Services/ServiceBase.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <param name="id">id</param>
+        /// <param name="field">названия поля, с возможным исключением</param>
+        /// <param name="cancellationToken"> токен отмены</param>
+        /// <typeparam name="TEntity">Дженерик типа</typeparam>
```
В комментах для дженерика надо указывать больше конкретики, например "Тип domain сущности" или "Тип репозитория"

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,113 @@
+public class CourseService : ServiceBase, ICourseService
+{
+    private readonly IMapper _mapper;
+    private readonly IStudentUnitOfWork _unitOfWork;
```
Зачем тебе здесь _unitOfWork, если он уже есть в ServiceBase ? Поправить во всех сервисах

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Interfaces/Services/ServiceBase.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <typeparam name="TEntity">Дженерик типа</typeparam>
+        /// <returns>При валидных данных и наличии модели в бд, возвращает эту модель</returns>
+        /// <exception cref="EntityNotFoundException{TEntity}"></exception>
+        protected async Task<TEntity> GetEntityByIdAsync<TEntity>(Guid id, string field, CancellationToken cancellationToken) where TEntity : BaseEntity
```
Можно обойтись и без передачи field (paramName), его можно вытащить и так, например nameof(BaseEntity.Id)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,121 @@
+        var educator = _mapper.Map<EducatorCreateDto, Educator>(educatorDto);
+        await _unitOfWork.EducatorRepository.AddAsync(educator, cancellationToken);
+        
+        
```
Тут лишняя пустая строка

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,121 @@
+        var educator = await GetEntityByIdAsync<Educator>(educatorDto.Id, "id", cancellationToken);
+        
+        educator.Update(
+            new FullName(educator.FullName.FirstName, educator.FullName.Surname, educator.FullName.Patronymic),
```
А может надо брать новые данные из educatorDto, а не из educator? )0)

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,121 @@
+    /// <returns>возвращает дто преподавателя, полученного по id</returns>
+    public async Task<EducatorResponseDto> GetByIdAsync(Guid id, CancellationToken cancellationToken)
+    {
+        Guard.Against.Default(id);
```
Да наверное лучше перенести эту проверку в GetEntityByIdAsync, чтобы каждый раз не писать

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,99 @@
+        
+        var enrollment = _mapper.Map<EnrollmentCreateDto, Enrollment>(enrollmentCreateDto);
+        
+        await GetEntityByIdAsync<Course>(enrollment.CourseId, "id", cancellationToken);
```
А есть ли нам смысл выполнять полное получение объекта БД только для того, что проверить его на существование ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,99 @@
+    /// </summary>
+    /// <param name="cancellationToken">токен отмены</param>
+    /// <returns>возвращает все зачисления, что удалось получить через бд</returns>
+    public async Task<List<EnrollmentResponseDto>> GetListStudentsAsync(CancellationToken cancellationToken)
```
GetListStudents? Название метода поправь

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/StudentService.cs:
```diff
@@ -0,0 +1,120 @@
+    /// </summary>
+    /// <param name="cancellationToken">токен отмены</param>
+    /// <returns>возвращает дто всех студентов</returns>
+    public async Task<List<StudentResponseDto>> GetInfoStudentsAsync(CancellationToken cancellationToken)
```
Везде для таких методов использвал GetList, а теперь какой-то Info. Используй уже одни и те же обозначения для тех или иных действий.

1nsanov commented:
Поправить документацию в сервисах, как писал в телеге

MrazotaIzTirasa committed:
Eduh 2.1
Правка багов, разработка
Удалил BaseService, попправил комментарии, исправил автомаппинг и др

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Mapping/EducatorMappingProfile.cs:
```diff
@@ -0,0 +1,51 @@
+                .ForMember(dest => dest.Patronymic, opt => opt.MapFrom(src => src.FullName.Patronymic))
+                .ForMember(dest => dest.Phone, opt => opt.MapFrom(src => src.Phone.Value))
+                .ForMember(dest => dest.YearsExperience, opt => opt.MapFrom(src => src.YearsExperience))
+                .ForMember(dest => dest.DateEmployment, opt => opt.MapFrom(src => src.DateEmployment));
```
Для YearsExperience и DateEmployment ForMember можно убрать

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Primitives/RegexPatterns.cs:
```diff
@@ -0,0 +1,24 @@
+    public const string PmrPhonePattern = @"^\+373(2\d{7}|[67]\d{7})$";
+        
+    /// <summary>
+    /// Паттерн для проверки почтового ящика
```
Почтовый ящик это на улице коробка, а у нас эл. почта )

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,100 @@
+    public async Task<CourseResponseDto> GetByIdAsync(Guid id, CancellationToken cancellationToken)
+    {
+        Guard.Against.Default(id);
+        var course = await this.GetByIdOrThrowAsync(id, cancellationToken);
```
Зачем this ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,100 @@
+    private async Task<Course> GetByIdOrThrowAsync(Guid id, CancellationToken cancellationToken)
+    {
+        var repository = _unitOfWork.GetRepository<ICourseRepository>();
+        var entity = await repository.GetByIdAsync(id, cancellationToken);
```
Уже можно написать _unitOfWork.CoursesRepository.GetByIdAsync(id, cancellationToken); и убрать сверху получение репы из DI. Исправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/CourseService.cs:
```diff
@@ -0,0 +1,100 @@
+        var entity = await repository.GetByIdAsync(id, cancellationToken);
+        if (entity == null)
+        {
+            throw new EntityNotFoundException<Course>(nameof(BaseEntity.Id), id);
```
Тут тоже пиши явно, nameof(Course.Id), поправить везде

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EducatorService.cs:
```diff
@@ -0,0 +1,107 @@
+        var entity = await repository.GetByIdAsync(id, cancellationToken);
+        if (entity == null)
+        {
+            throw new EntityNotFoundException<Course>(nameof(BaseEntity.Id), id);
```
EntityNotFoundException<Educator> должно же быть

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,97 @@
+    public EnrollmentService(IMapper mapper, IStudentUnitOfWork unitOfWork)
+    {
+        _mapper = Guard.Against.Null(mapper);
+        _unitOfWork = unitOfWork;
```
Схуяли убрал Guard.Against.Null ?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,97 @@
+        var isEnrollment = await _unitOfWork.CoursesRepository.ExistAsync(enrollment.CourseId, cancellationToken);
+        if (!isEnrollment)
+        {
+            throw new EntityNotFoundException<Course>(nameof(BaseEntity.Id), enrollment.CourseId);
```
Дженерик тип для ошибки не верный

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,97 @@
+        var isStudent = await _unitOfWork.StudentRepository.ExistAsync(enrollment.StudentId, cancellationToken);
+        if (!isStudent)
+        {
+            throw new EntityNotFoundException<Course>(nameof(BaseEntity.Id), enrollment.StudentId);
```
Дженерик тип для ошибки не верный

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/EnrollmentService.cs:
```diff
@@ -0,0 +1,97 @@
+    {
+        var enrollments = await _unitOfWork.EnrollmentRepository.GetAllAsync(cancellationToken);
+        
+        return _mapper.Map<List<EnrollmentResponseDto>>(enrollments).ToList();
```
Че, а зачем .ToList(); в конце?

1nsanov commented on src/Services/StudentService/Application/EduHub.StudentService.Application.Services/Services/StudentService.cs:
```diff
@@ -0,0 +1,107 @@
+    public StudentService(IMapper mapper, IStudentUnitOfWork unitOfWork)
+    {
+        _mapper = Guard.Against.Null(mapper);
+        _unitOfWork = unitOfWork;
```
Куда делся Guard.Against.Null ?

MrazotaIzTirasa committed:
EDUH 2.1
Косметические правки
